src/Utils.cc

```diff
@@ -563,4 +574,271 @@ thread_type_t fbgemmGetThreadPartition(
   return th_info;
 }
 
+namespace {
+
+// implementation taken from pytorch/c10/util/llvmMathExtras.h
+template <typename T>
+size_t count_leading_zeros(T val) {
+  if (!val)
+    return std::numeric_limits<T>::digits;
+
+  size_t zero_bits = 0;
+  for (T shift = std::numeric_limits<T>::digits >> 1; shift; shift >>= 1) {
+    T tmp = val >> shift;
+    if (tmp)
+      val = tmp;
+    else
+      zero_bits |= shift;
+  }
+  return zero_bits;
+}
+
+// histogram size per thread
+constexpr int RDX_HIST_SIZE = 256;
+
+#define COMBINE_PREFIX_SUM_IN_RANGE(                            \
```

Comment:
Can this be an actual function? I'd suggest one that returns `void` and takes `offset` as an input instead of `prev_sum`.

Reply:
Yeah, good idea. I created a new function with `offset` as an in-out argument.